### ISSUE TITLE
Uncaught error handler is not threadsafe in Fuchsia embedder

### ISSUE THREAD

**Issue by cbracken** [flutter/flutter#34726]
In the `flutter_runner` embedder for Fuchsia, we landed mitigation to reduce the likelihood of a crash in the uncaught error handler when an uncaught exception is thrown immediately at/during shutdown. Related patch is https://fuchsia-review.googlesource.com/c/topaz/+/292831.

The error handling as originally landed was unsafe. Specifically, in the case where the unhandled error handler was triggered during shutdown, there was a race condition which could cause a crash in the following scenario:
1. `Runner::OnApplicationTerminate()` is triggered, which posts a task to the application's platform thread will free the `Application` instance and terminate the platform thread.
2. Before that task is serviced, the unhandled error handler is called (by `hooks.dart` -> `window.cc` -> `ui_dart_state.cc`) on the UI thread.
3. The kill task is serviced and the `Application` dtor and `Thread::Quit()` are called, terminating the platform thread.
4. The unhandled error handler attempts to post a task to the platform thread, whose thread was killed in step 3. This triggers a crash.

Fixing this requires a mechanism for the message loop to know that the associated thread has been terminated out from under it.

This patch adds mitigation for this scenario, but remains non-threadsafe/racy. We pass the unhandled error handler a weak pointer to the `Application` and check it before posting a task to the platform thread. This has two issues:
1. `WeakPtr` isn't threadsafe (nor is it intended to be), and assumes that all operations occur on a single thread. We're checking its value (which is mutated on the platform thread) on the UI thread without synchronization.
2. Even with a guarantee that the `WeakPtr` state were synchronized, there's a window between when we check the weak pointer and when we post to the platform thread in which application shutdown and thread destruction may occur.

This unsafe mitigation was landed in order to unblock a high priority bug ([FL-256](https://fuchsia.atlassian.net/browse/FL-256)) on a short schedule, and a proper refactoring is required to make this properly threadsafe.

/cc @cbracken @chinmaygarde

**Comment by zoeyfan** [comment]
@cbracken I am triaging our issue database, and noticed this bug from earlier this year. Is this still relevant? Can we close it now?

**Comment by gw280** [comment]
@zoeyfan This is still relevant but it is very rare and fairly low priority. It looks like it would require substantial work to resolve.
